src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -215,18 +200,25 @@ private unsafe Socket(SafeSocketHandle handle, bool loadPropertiesFromHandle)
                     try
                     {
                         socketAddress ??= new Internals.SocketAddress(_addressFamily, buffer);
-                        if (SocketPal.GetPeerName(_handle, socketAddress.Buffer, ref socketAddress.InternalSize) != SocketError.Success)
-                        {
-                            return;
-                        }
 
-                        if (socketAddress.InternalSize > socketAddress.Buffer.Length)
+                        SocketError result = SocketPal.GetPeerName(_handle, socketAddress.Buffer, ref socketAddress.InternalSize);
+                        if ((RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && result == SocketError.Fault) ||
+                            (socketAddress.InternalSize > socketAddress.Buffer.Length))
```

Comment:
The code above that calls `GetSockName` is subject to the same behavior on Windows. It doesn't get into that situation, because the buffer used is `512` bytes, which is sufficiently large. We should use the same buffer here to start with, and not a slice.

On Linux `sockaddr_storage` is `128`, so that is the larges buffer one may need. I don't know what the value is on Windows? Maybe we can capture the cross-platform max in a `const`, and assert it in PAL? Or add a PAL function to retrieve it?

In https://github.com/dotnet/runtime/pull/34727 some code was also added to handle the case where the provided buffer was too small. I don't think we get in that situation, but the enlarged buffer isn't actually larger:
```
                if (bufferLength > buffer.Length)
                {
                    buffer = new byte[buffer.Length];
```
The new buffer allocated should be `bufferLength` length.

Reply:
> The new buffer allocated should be bufferLength length.

Ugh, yes.  Silly period.

Reply:
On Linux we can reallocate. On OSX & Windows sockaddr_un is largest AF AFAIK. How about putting 128 const to Socket.cs and adding Debug.Assert after GetSockName and GetPeerName @stephentoub ? 
Or do we want to keep the 2 stage approach to keep is small as possible? 128 seems large for IPv4 but I'm not sure how much this matter as this API is probably not the main people use to deal with Sockets.

Reply:
> On Linux we can reallocate. On OSX & Windows sockaddr_un is largest AF AFAIK. How about putting 128 const to Socket.cs and adding Debug.Assert after GetSockName and GetPeerName @stephentoub ?

A value that is max `sizeof(sockaddr_storage)` across all platforms.
`SocketAddress` may be more appropriate place for this.
With a `static_assert` in pal that verifies this value is `<= sizeof(sockaddr_storage)`.

There may be other issues with `GetSockName` and `GetPeerName` for unix addresses when used for `{Local/Remote}EndPoint`. I think the code assumes that `_rightEndPoint.Serialize()` returns a `SocketAddress` that is large enough, which may not be true if `_rightEndPoint` is a small path.

> Or do we want to keep the 2 stage approach to keep is small as possible? 128 seems large for IPv4 but I'm not sure how much this matter as this API is probably not the main people use to deal with Sockets.

Can we `stackalloc` and then do `Slice(0, bufferLength).ToArray()`?

Reply:
the problem goes deeper. I modified some UDS tests to have different  length between client and server and I'm getting asserts in networking PAL. GetRandomNonExistingFilePath() always conveniently returns same length for both.

Reply:
I added const as @tmds suggested. It seems like we do not need to slice() for IP4/6. The pall can happily take more bytes then needed. Is I use same buffer for remote EP. 
I bump to problem than  Accept & RemoteEndPoint would fail if remote address is longer than local one. That was never problem before as IP protocols have fixed address length but UDS does not.